### ISSUE TITLE
OOM

### ISSUE THREAD

**Issue by mschwartz** [rvrolyk/SleepNumberController#11]
I have been using the app for a few weeks now.  I have a few issues with it, but I want to mention the big two here.

First, ever since I started using the app, my Hubitat c5 reports it’s running low on memory a few days after a reboot.  The warning appears where the notices for updates to the platform appear, along with the suggestion to reboot.  If I don’t reboot the hub, it eventually becomes completely unresponsive and I have to pull the plug and reinsert it to power cycle the hub.

I was polling day and night at 1.  Today I adjusted the setting to poll every 10 during day and night hours that I set.

second, the presence detection does not seem to work at all.  I can change it in the app’s settings using the arrived and departed “buttons “.

Finally, the level value appears to be the current level of the bed.  Like it starts at 50, I change the sleepNumber to 75 with the remote or app and level eventually becomes 75 (as the bed inflates).  The app interface has a setLevel for some reason that I don’t understand.

if you have questions about my set up (like bed is a C series with only SleepIQ data and lef/right baffles, no other adjustments), I would like to help!

<img width="892" alt="image" src="https://github.com/rvrolyk/SleepNumberController/assets/329637/6bcc9457-fca7-4699-8f21-2d668cb7829b">

**Comment by rvrolyk** [comment]
I haven’t heard any other reports of abnormal memory usage. It is a big app and polling 1440 times a day will increase cpu usage but the memory footprint should remain stable. Can you provide the app activity table from the hub (found on logs page)?  If you have sleep data collection enabled that will also increase memory so if you aren’t using it, you might try disabling that. 

I’m surprised presence doesn’t work, it’s usually the only stables thing across the beds. Can you enable debug logs (on the app page itself) and then click poll on one if your parent bed devices and share that output here? A screenshot is usually easier to read than copy/paste. 

Level indicates different things depending on the device. It sounds like you are referring to the parent device and in that case it’s the mattress pressure (eg your sleep number). The setLevel command will change that value. 

On/off is described in the help — 
>Set Bed Preset*: sets the bed to a preset level. If the device type is presence, head or foot, then on will set enable the preset set as a preference (in this way, you can have 3 preferred preset levels via voice) and off will set the bed to flat.

If you look in the device preferences you’ll see a “bed preset level for on” option and that correlates to the on value.

**Comment by mschwartz** [comment]
I’m still having the OOM issue.  It doesn’t seem to matter what polling options I use - longer time between polls extends the time until OOM.  It happens after about 3 days with my current settings.

I dug into the code, but my knowledge of groovy and how it garbage collects is lacking.

the code is impressive, and it’s clearly a lot of work.

my settings are 30 minutes during the day, 2 minutes overnight.

**Comment by rvrolyk** [comment]
That's really curious as I haven't heard of any persistent reports of OOM.
Are you running the latest version of Hubitat OS?   Can you share what your
app and device settings are?  Maybe you're using a feature that isn't as
often exercised so this has gone unnoticed?   Do you see any errors or
warnings in the log?
Obviously a single sample but I don't reboot my hub except for updates and
I poll every minute overnight and every 30 during the day.

**Comment by mschwartz** [comment]
<img width="1357" alt="image" src="https://github.com/rvrolyk/SleepNumberController/assets/329637/fc9495d1-8773-4476-9bea-cb34b60484d2">

**Comment by mschwartz** [comment]
Hub is up to date, FWIW.  I’ll post all the screens.

**Comment by mschwartz** [comment]
<img width="1120" alt="image" src="https://github.com/rvrolyk/SleepNumberController/assets/329637/604e1149-8f27-4cab-ad43-f3be24c6958c">

**Comment by mschwartz** [comment]
<img width="1125" alt="image" src="https://github.com/rvrolyk/SleepNumberController/assets/329637/88c542ad-912e-4e8b-8db0-a52f1e2be857">

**Comment by mschwartz** [comment]
I found this app to be a bit touchy.  If I change things, it gets messed up to the point where I end up deleting and reinstalling the app.

The bed has just left/right side, no other bells and whistles.  All I’m expecting is presence and sleepNumber/level and SleepIQ stats.

**Comment by mschwartz** [comment]
<img width="1125" alt="image" src="https://github.com/rvrolyk/SleepNumberController/assets/329637/3d8dd7b5-194f-4e5e-9b55-b7d762916274">

**Comment by mschwartz** [comment]
It’s going to take several screen shots to get you the individual device details and/or logs.  Let me know if you need those.

**Comment by rvrolyk** [comment]
I'm just thinking a screenshot of each device preferences section.  I want
to see how your setup compares to mine so I can try to reproduce your issue

**Comment by mschwartz** [comment]
<img width="1123" alt="image" src="https://github.com/rvrolyk/SleepNumberController/assets/329637/c794eb00-59d2-4782-9952-3da542a93ffb">

**Comment by mschwartz** [comment]
<img width="1121" alt="image" src="https://github.com/rvrolyk/SleepNumberController/assets/329637/9963e764-621f-450f-b35d-8c188543701b">

**Comment by mschwartz** [comment]
Note that I am using this with Maker API, which might be different than what you do.

The data looks good.  It just runs out of memory.

My analysis leads to this app.  I never had out of memory before installing it and setting it up. It could be something to do with a later version of the Hubitat firmware.

Plus as I adjusted the poll time overnight from 1 to 2, the time between OOM got a few hours longer.

It’s possible that my hub has much more limited memory than yours and the JVM (groovy) garbage collector gets more time to run on your hub.

All this is my pure speculation.

**Comment by rvrolyk** [comment]
Thanks. That looks pretty similar to mine I think and you don’t appear to
have sleep data collection enabled on either so actual state should be low.

When you have a minute, can you grab a snapshot of your app statistics from
the logs page?  I’m wondering if it looks similar to mine or is somehow
very different.

**Comment by mschwartz** [comment]
I am getting sleep data reported.  It changes daily.

<img width="1358" alt="image" src="https://github.com/rvrolyk/SleepNumberController/assets/329637/9528ebde-97b3-48d0-af66-14e95a9512f1">

**Comment by mschwartz** [comment]
<img width="1127" alt="image" src="https://github.com/rvrolyk/SleepNumberController/assets/329637/39f430af-c66c-4084-93f0-12bd3caafb85">

**Comment by mschwartz** [comment]
Unfortunately, I rebooted the hub about 20 minutes ago.  The data may not be ideal.

**Comment by rvrolyk** [comment]
Oh that’s interesting since it’s turned off on the device. I’ll have to
check how I coded that… I assume you want sleep data?

**Comment by rvrolyk** [comment]
Yeah it probably hasn’t done much. Data after the night would be best I
think.

**Comment by mschwartz** [comment]
<img width="1123" alt="image" src="https://github.com/rvrolyk/SleepNumberController/assets/329637/efc6a3ce-2401-4dde-b82d-955bed69adea">

**Comment by mschwartz** [comment]
Z wave poller app is working against the Bathroom Light device…

**Comment by mschwartz** [comment]
I asked earlier about sleep data.  I think these settings are consistent with what you told me at the time.  I was manually forcing the poll…

**Comment by mschwartz** [comment]
Another bit of information.  I have seen a lot of these unauthorized log messages when I try to diagnose things.

<img width="1123" alt="image" src="https://github.com/rvrolyk/SleepNumberController/assets/329637/7dab92e4-2eb7-4954-87d4-b1c283d92569">

**Comment by mschwartz** [comment]
App 424 is this one…. The app reports online, even with the messages.

**Comment by mschwartz** [comment]
I'm looking at your code.  In error conditions, you call functions recursively, which can grow the stack.

https://github.com/rvrolyk/SleepNumberController/blob/master/SleepNumberController_App.groovy#L1874

httpRequest calls login calls fail calls httpRequest calls login ....

Now, I'm not surety lifecycle of how these routines are being called.

Just a thought with only a brief review of the code...  Have the login function return if already logged in and return or throw an error if login fails.  If it fails, go offline.  

In the loop where you poll, periodically, login (if fail, set offline, done polling), then poll each thing.

In httpRequest, if fails, throw an error, and stop the polling right there.

Maybe I'm full of it :)

**Comment by mschwartz** [comment]
![2023-07-07 at 4 19 PM](https://github.com/rvrolyk/SleepNumberController/assets/329637/e14a7e9e-0075-4329-9d21-3dd6c8d9be01)

**Comment by rvrolyk** [comment]
It should trigger a token refresh and then if it fails again, fail the
entire thing. I’ll take another look though and make sure it’s actually
working that way.

**Comment by mschwartz** [comment]
![2023-07-07 at 4 30 PM](https://github.com/rvrolyk/SleepNumberController/assets/329637/e8091ca6-3447-4d7c-a117-4abf113aac29)

**Comment by mschwartz** [comment]
Looks like every 30 minutes.

**Comment by rvrolyk** [comment]
That should match your polling. I can’t recall how long the auth token
lasts but it looks like it’s expiring before the 30 minute mark so the
first request triggers a refresh and the second request succeeds (with the
fresh token)
